### PR TITLE
WIP: Add clightning backup plugin

### PR DESCRIPTION
**Pull request by jonasnick**

### CODE REVIEW

**Comment by jonasnick** [comment]
> TODO: clightning fails with "Could not find backup.lock in the lightning-dir". Yet, backup.lock clearly exists in the lightning-dir.

Was able to fix this by using `strace` which showed that the plugin's working directory is actually `dataDir/bitcoin`

Now need to figure out a good way to give clightning (and clightning-backup-init) permissions to write into an arbitrary directory and not just clightnings dataDir.

File: modules/clightning-plugins/backup.nix [hunk 1]
```diff
@@ -0,0 +1,51 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.services.clightning.plugins.backup;
+  nbLib = config.nix-bitcoin.lib;
+  nbPkgs = config.nix-bitcoin.pkgs;
+  clightningCfg = config.services.clightning;
```
Comment (nixbitcoin) [inline modules/clightning-plugins/backup.nix:8]: We call this `clightning` in other packages...

File: modules/clightning-plugins/backup.nix [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      default = "";
+      description = "Location of the backup";
+      example = "file:///mnt/external/location/file.bkp";
+    };
+    inherit (nbLib) enforceTor;
+  };
+  config = mkIf cfg.enable {
+    assertions = [
+      { assertion = cfg.backendURL != "";
+        message = "You must set system.services.clightning.plugins.backup.backendURL";
```
Comment (nixbitcoin) [inline modules/clightning-plugins/backup.nix:25]: Shouldn't this be `config.services.clightning.plugins.backup.backendURL`?

File: modules/clightning-plugins/backup.nix [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+          Type = "oneshot";
+          RemainAfterExit = true;
+          User = clightningCfg.user;
+        };
+        script = ''
+          backendURL="${cfg.backendURL}"
+          backupLock="${clightningCfg.networkDir}/backup.lock"
+          if [[ ! -e $backupLock || $backendURL != $(cat $backupLock | ${pkgs.jq}/bin/jq -r .backend_url) ]]; then
+            umask u=rw,go=
+            ${backupPkg}/backup-cli init --lightning-dir "${clightningCfg.networkDir}" ${cfg.backendURL}
```
Comment (nixbitcoin) [inline modules/clightning-plugins/backup.nix:46]: Can you add `${backupPkg}/backup-cli` to the system packages for restores?

**Comment by seberm** [comment]
@jonasnick 
Please also do not forget to add an information about `backup` plugin into the README file (https://github.com/fort-nix/nix-bitcoin/blob/3f844c06f08bee8f3da0141fcba8d76ca95ef992/README.md?plain=1#L58).

Thanks for your work!

File: modules/clightning-plugins/backup.nix [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        before = [ "clightning.service" ];
+        serviceConfig = {
+          Type = "oneshot";
+          RemainAfterExit = true;
+          User = clightningCfg.user;
+        };
+        script = ''
+          backendURL="${cfg.backendURL}"
+          backupLock="${clightningCfg.networkDir}/backup.lock"
+          if [[ ! -e $backupLock || $backendURL != $(cat $backupLock | ${pkgs.jq}/bin/jq -r .backend_url) ]]; then
```
Comment (nixbitcoin) [inline modules/clightning-plugins/backup.nix:44]: What if backups get deleted upstream?

**Comment by nixbitcoin** [comment]
Tested the backup plugin in production with a `rclone`/`webdav`/`nextcloud` directory mounted by a systemd service in `clightning.dataDir`. A lot of manual fiddling but it works.

**Comment by erikarvstedt** [comment]
Superseded by https://github.com/fort-nix/nix-bitcoin/pull/415.

**Pull request closed without merging**